### ISSUE TITLE
Deprecation issue : the behavior of changed inside after callbacks...

### ISSUE THREAD

**Issue by zedalaye** [rails/rails#29744]
### Steps to reproduce
Update a boolean attribute

### Expected behavior
The boolean attribute is updated

### Actual behavior
I get this pile of deprecation warnings:
```
DEPRECATION WARNING: The behavior of `changed_attributes` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_changes.transform_values(&:first)` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `changes` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_changes` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `changed` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_changes.keys` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `attribute_change` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `attribute_changed?` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute?` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `attribute_change` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `attribute_changed?` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute?` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `changed_attributes` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_changes.transform_values(&:first)` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `changes` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_changes` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `changed` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_changes.keys` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `attribute_change` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `attribute_changed?` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute?` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `attribute_change` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
DEPRECATION WARNING: The behavior of `attribute_changed?` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute?` instead. (called from update_acl at /home/pierre/Projets/rails/rezolution/app/models/concerns/grantable.rb:40)
```

### System configuration
**Rails version**: 5.1.2

**Ruby version**: 2.4.1

Here is my "Grantable" concern (look : no callback at all)

```ruby
module Grantable
  extend ActiveSupport::Concern

  included do

  end

  module ClassMethods
    def allowed(profile_types)
      if profile_types.empty?
        none
      else
        profile_types = [ profile_types ] unless profile_types.is_a?(Array)
        condition = profile_types.map { |profile_type| "(acl->>'#{profile_type}'='true')" }.join(' OR ')
        where(condition)
      end
    end
  end

  def allow_for(profile_type)
    update_acl(profile_type, true)
  end

  def deny_for(profile_type)
    update_acl(profile_type, false)
  end

  def allowed?(profile_type)
    !!(self.acl and self.acl[profile_type.to_s])
  end

  def toggle_acl_for(profile_type)
    update_acl(profile_type, !allowed?(profile_type))
  end

  private

  def update_acl(profile_type, allowed)
    # HERE
    self.update_attribute(:acl, (self.acl || {}).merge( { profile_type => allowed } ))
  end

end
```

The corresponding controller action:

```ruby
module Library
  class DocumentsController < BaseController
    before_action :check_company_admin
    before_action :set_folder, only: [ :create ]
    before_action :set_file, only: [ :edit, :update, :destroy, :toggle_acl ]

    # ... skipped

    def toggle_acl
      if @file.toggle_acl_for(params[:profile_type])
        head :ok
      else
        head :unprocessable_entity
      end
    end

    private

    def set_folder
      @folder = @current_company.folders.find(params[:folder_id]) if params[:folder_id]
    end

    def set_file
      @file = @current_company.documents.find(params[:id])
    end

    def document_params
      params.require(:document).permit([ :contents, :name ])
    end

  end
end
```

**Comment by zedalaye** [comment]
Please note that the ACL is stored in a json column.

```ruby
create_table :documents do |t|
  t.belongs_to :company # + foreign key
  t.belongs_to :folder # + foreign key
  # paperclip stuff
  t.json acl
  t.timestamps
end
```

**Comment by matthewd** [comment]
The deprecation message filters gem lines, so if the callbacks are defined in a gem, it will misleadingly blame the application line that triggered the original save.

**Comment by sgrif** [comment]
As mentioned, there is a gem which is defining callbacks which are causing the deprecation warning. I'm going to close this, as there's nothing actionable for us to do here.

**Comment by zedalaye** [comment]
Thank you @matthewd and @sgrif for pointing my attention on a gem defining that callback.

For those who care, in my case, it was due to ClosureTree gem. Fixed in their v6.6.0 by https://github.com/ClosureTree/closure_tree/pull/262
